File: src/jedimem/repo.py
"""Git access.

Every write path here is deliberately conservative, because the measurements in
docs/research/04-git-format.md showed the obvious approaches lose data:

  * `git add` + `git commit` from a background process lost 13 of 20 memories to
    .git/index.lock contention, silently.
  * A lock-free commit onto the checked-out branch was then silently reverted by
    the user's next ordinary `git commit` (their index was stale).

So: we never touch the index, never advance refs/heads/*, and never write to the
working tree except when a human explicitly asks (promote/compile).
"""
from __future__ import annotations

import os
import pathlib
import subprocess
import tempfile

STAGING_REF = "refs/jedimem/log"


class GitError(RuntimeError):
    pass


def _run(args, cwd=None, check=True, stdin=None):
    p = subprocess.run(["git"] + args, cwd=cwd, capture_output=True, text=True,
                       input=stdin)
    if check and p.returncode != 0:
        raise GitError(f"git {' '.join(args)}: {p.stderr.strip()}")
    return p.stdout.strip(), p.returncode


def is_repo(cwd=None) -> bool:
    _, rc = _run(["rev-parse", "--git-dir"], cwd=cwd, check=False)
    return rc == 0


def toplevel(cwd=None) -> pathlib.Path:
    out, rc = _run(["rev-parse", "--show-toplevel"], cwd=cwd, check=False)
    if rc != 0:
        raise GitError("not inside a git repository")
    return pathlib.Path(out)


def git_dir(cwd=None) -> pathlib.Path:
    """Per-worktree. The right home for per-checkout state."""
    out, _ = _run(["rev-parse", "--absolute-git-dir"], cwd=cwd)
    return pathlib.Path(out)


def common_dir(cwd=None) -> pathlib.Path:
    """Shared by all worktrees of a repo. The local scope key."""
    out, _ = _run(["rev-parse", "--path-format=absolute", "--git-common-dir"], cwd=cwd)
    return pathlib.Path(out)


def worktrees(cwd=None) -> list:
    out, _ = _run(["worktree", "list", "--porcelain"], cwd=cwd)
    trees, cur = [], {}
    for line in out.splitlines():
        if not line.strip():
            if cur:
                trees.append(cur); cur = {}
            continue
        k, _, v = line.partition(" ")
        cur[k] = v
    if cur:
        trees.append(cur)
    return trees


def head_commit(cwd=None) -> str:
    out, rc = _run(["rev-parse", "HEAD"], cwd=cwd, check=False)
    return out if rc == 0 else ""


def current_branch(cwd=None) -> str:
    out, rc = _run(["rev-parse", "--abbrev-ref", "HEAD"], cwd=cwd, check=False)
    return out if rc == 0 else ""


def normalized_remote(cwd=None) -> str:
    """Strip the incidental differences between spellings of the same remote."""
    url, rc = _run(["remote", "get-url", "origin"], cwd=cwd, check=False)
    if rc != 0 or not url:
        return ""
    u = url.strip()
    for pre in ("git+ssh://", "ssh://", "https://", "http://", "git://"):
        if u.startswith(pre):
            u = u[len(pre):]
    if "@" in u.split("/")[0]:
        u = u.split("@", 1)[1]
    u = u.replace(":", "/", 1) if "/" not in u.split(":")[0] else u
    if u.endswith(".git"):
        u = u[:-4]
    return u.rstrip("/").lower()


def root_commit(cwd=None) -> str:
    """NOT a stable identity: a --depth 1 clone yields a different root."""
    out, rc = _run(["rev-list", "--max-parents=0", "HEAD"], cwd=cwd, check=False)
    return out.splitlines()[0] if rc == 0 and out else ""


# ---------------------------------------------------------------- staging ref

def _write_blob(content: str, cwd=None) -> str:
    out, _ = _run(["hash-object", "-w", "--stdin"], cwd=cwd, stdin=content)
    return out


def ref_exists(ref=STAGING_REF, cwd=None) -> str:
    out, rc = _run(["rev-parse", "--verify", "-q", ref], cwd=cwd, check=False)
    return out if rc == 0 else ""


def stage_files(files: dict, message: str, ref=STAGING_REF, cwd=None,
                retries: int = 100) -> str:
    """Commit {path: content} onto a side ref without touching index/HEAD/worktree.

    Compare-and-swap on the ref, retrying on contention. The parent MUST be read
    before the tree is built: doing it the other way round passes the CAS while
    committing a stale tree, which silently drops other writers' work.
    """
    blobs = {p: _write_blob(c, cwd=cwd) for p, c in files.items()}
    for _ in range(retries):
        parent = ref_exists(ref, cwd=cwd)
        # mktemp NAME ONLY: git refuses an existing zero-byte index file.
        fd, idx = tempfile.mkstemp(prefix="jedimem-idx-")
        os.close(fd); os.unlink(idx)
        env = dict(os.environ, GIT_INDEX_FILE=idx)
        try:
            if parent:
                subprocess.run(["git", "read-tree", parent], cwd=cwd, env=env,
                               capture_output=True, check=True)
            for path, blob in blobs.items():
                subprocess.run(["git", "update-index", "--add", "--cacheinfo",
                                f"100644,{blob},{path}"], cwd=cwd, env=env,
                               capture_output=True, check=True)
            tree = subprocess.run(["git", "write-tree"], cwd=cwd, env=env,
                                  capture_output=True, text=True, check=True).stdout.strip()
        finally:
            if os.path.exists(idx):
                os.unlink(idx)

        args = ["commit-tree", tree, "-m", message] + (["-p", parent] if parent else [])
        commit, _ = _run(args, cwd=cwd)
        _, rc = _run(["update-ref", ref, commit, parent or ""], cwd=cwd, check=False)
        if rc == 0:
            return commit
    raise GitError(f"could not update {ref} after {retries} attempts")


def staged_files(ref=STAGING_REF, cwd=None, prefix="") -> list:
    if not ref_exists(ref, cwd=cwd):
        return []
    out, _ = _run(["ls-tree", "-r", "--name-only", ref], cwd=cwd)
    return [p for p in out.splitlines() if p.startswith(prefix)]


def staged_content(path: str, ref=STAGING_REF, cwd=None) -> str:
    out, _ = _run(["show", f"{ref}:{path}"], cwd=cwd)
    return out


def drop_staged(paths, ref=STAGING_REF, cwd=None, message="jedimem: clear staged") -> str:
    """Rewrite the staging ref without `paths` (after promote/reject)."""
    keep = {}
    for p in staged_files(ref=ref, cwd=cwd):
        if p not in paths:
            keep[p] = staged_content(p, ref=ref, cwd=cwd) + "\n"
    parent = ref_exists(ref, cwd=cwd)
    fd, idx = tempfile.mkstemp(prefix="jedimem-idx-"); os.close(fd); os.unlink(idx)
    env = dict(os.environ, GIT_INDEX_FILE=idx)
    try:
        for path, content in keep.items():
            blob = _write_blob(content, cwd=cwd)
            subprocess.run(["git", "update-index", "--add", "--cacheinfo",
                            f"100644,{blob},{path}"], cwd=cwd, env=env,
                           capture_output=True, check=True)
        tree = subprocess.run(["git", "write-tree"], cwd=cwd, env=env,
                              capture_output=True, text=True, check=True).stdout.strip() \
            if keep else subprocess.run(["git", "hash-object", "-w", "-t", "tree", "/dev/null"],
                                        cwd=cwd, capture_output=True, text=True,
                                        check=True).stdout.strip()
    finally:
        if os.path.exists(idx):
            os.unlink(idx)
    args = ["commit-tree", tree, "-m", message] + (["-p", parent] if parent else [])
    commit, _ = _run(args, cwd=cwd)
    _run(["update-ref", ref, commit, parent or ""], cwd=cwd)
    return commit

File: src/repo.rs
//! Git access.
//!
//! Every write path here is deliberately conservative, because the measurements
//! in docs/research/04-git-format.md showed the obvious approaches lose data:
//!
//!   * `git add` + `git commit` from a background process lost 13 of 20
//!     memories to .git/index.lock contention, silently.
//!   * A lock-free commit onto the checked-out branch was then silently
//!     reverted by the user's next ordinary `git commit` (their index was
//!     stale), leaving spurious `D` entries in their status.
//!
//! So: we never touch the index, never advance refs/heads/*, and never write to
//! the working tree except when a human explicitly asks.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

pub const STAGING_REF: &str = "refs/jedimem/log";

#[derive(Debug)]
pub struct GitError(pub String);

impl std::fmt::Display for GitError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}
impl std::error::Error for GitError {}

pub type R<T> = Result<T, GitError>;

fn base(cwd: Option<&Path>) -> Command {
    let mut c = Command::new("git");
    if let Some(d) = cwd {
        c.current_dir(d);
    }
    c
}

/// Run git, returning (stdout, exit_ok).
pub fn run(args: &[&str], cwd: Option<&Path>) -> R<(String, bool)> {
    let out = base(cwd)
        .args(args)
        .output()
        .map_err(|e| GitError(format!("git not available: {}", e)))?;
    Ok((
        String::from_utf8_lossy(&out.stdout).trim().to_string(),
        out.status.success(),
    ))
}

pub fn check(args: &[&str], cwd: Option<&Path>) -> R<String> {
    let out = base(cwd)
        .args(args)
        .output()
        .map_err(|e| GitError(format!("git not available: {}", e)))?;
    if !out.status.success() {
        return Err(GitError(format!(
            "git {}: {}",
            args.join(" "),
            String::from_utf8_lossy(&out.stderr).trim()
        )));
    }
    Ok(String::from_utf8_lossy(&out.stdout).trim().to_string())
}

pub fn toplevel(cwd: Option<&Path>) -> R<PathBuf> {
    let (out, ok) = run(&["rev-parse", "--show-toplevel"], cwd)?;
    if !ok {
        return Err(GitError("not inside a git repository".into()));
    }
    Ok(PathBuf::from(out))
}

/// Per-worktree. The right home for per-checkout state.
pub fn git_dir(cwd: Option<&Path>) -> R<PathBuf> {
    Ok(PathBuf::from(check(
        &["rev-parse", "--absolute-git-dir"],
        cwd,
    )?))
}

/// Shared by all worktrees of a repo. The local scope key.
pub fn common_dir(cwd: Option<&Path>) -> R<PathBuf> {
    Ok(PathBuf::from(check(
        &["rev-parse", "--path-format=absolute", "--git-common-dir"],
        cwd,
    )?))
}

pub fn worktrees(cwd: Option<&Path>) -> R<Vec<BTreeMap<String, String>>> {
    let out = check(&["worktree", "list", "--porcelain"], cwd)?;
    let mut trees = Vec::new();
    let mut cur: BTreeMap<String, String> = BTreeMap::new();
    for line in out.lines() {
        if line.trim().is_empty() {
            if !cur.is_empty() {
                trees.push(std::mem::take(&mut cur));
            }
            continue;
        }
        let (k, v) = line.split_once(' ').unwrap_or((line, ""));
        cur.insert(k.to_string(), v.to_string());
    }
    if !cur.is_empty() {
        trees.push(cur);
    }
    Ok(trees)
}

pub fn head_commit(cwd: Option<&Path>) -> String {
    run(&["rev-parse", "HEAD"], cwd)
        .map(|(s, ok)| if ok { s } else { String::new() })
        .unwrap_or_default()
}

pub fn current_branch(cwd: Option<&Path>) -> String {
    run(&["rev-parse", "--abbrev-ref", "HEAD"], cwd)
        .map(|(s, ok)| if ok { s } else { String::new() })
        .unwrap_or_default()
}

/// Strip the incidental differences between spellings of the same remote.
pub fn normalized_remote(cwd: Option<&Path>) -> String {
    let (url, ok) = match run(&["remote", "get-url", "origin"], cwd) {
        Ok(v) => v,
        Err(_) => return String::new(),
    };
    if !ok || url.is_empty() {
        return String::new();
    }
    let mut u = url.as_str();
    for pre in ["git+ssh://", "ssh://", "https://", "http://", "git://"] {
        if let Some(rest) = u.strip_prefix(pre) {
            u = rest;
        }
    }
    let mut s = u.to_string();
    if let Some(host) = s.split('/').next() {
        if host.contains('@') {
            s = s.split_once('@').map(|x| x.1).unwrap_or(&s).to_string();
        }
    }
    if !s.split(':').next().unwrap_or("").contains('/') {
        s = s.replacen(':', "/", 1);
    }
    if let Some(t) = s.strip_suffix(".git") {
        s = t.to_string();
    }
    s.trim_end_matches('/').to_lowercase()
}

/// NOT a stable identity: a --depth 1 clone yields a different root commit.
pub fn root_commit(cwd: Option<&Path>) -> String {
    match run(&["rev-list", "--max-parents=0", "HEAD"], cwd) {
        Ok((s, true)) => s.lines().next().unwrap_or("").to_string(),
        _ => String::new(),
    }
}

// ------------------------------------------------------------- staging ref

fn write_blob(content: &str, cwd: Option<&Path>) -> R<String> {
    use std::io::Write;
    let mut child = base(cwd)
        .args(["hash-object", "-w", "--stdin"])
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .map_err(|e| GitError(format!("git hash-object: {}", e)))?;
    child
        .stdin
        .as_mut()
        .ok_or_else(|| GitError("no stdin".into()))?
        .write_all(content.as_bytes())
        .map_err(|e| GitError(e.to_string()))?;
    let out = child
        .wait_with_output()
        .map_err(|e| GitError(e.to_string()))?;
    if !out.status.success() {
        return Err(GitError(String::from_utf8_lossy(&out.stderr).into_owned()));
    }
    Ok(String::from_utf8_lossy(&out.stdout).trim().to_string())
}

pub fn ref_exists(r: &str, cwd: Option<&Path>) -> String {
    match run(&["rev-parse", "--verify", "-q", r], cwd) {
        Ok((s, true)) => s,
        _ => String::new(),
    }
}

/// A temporary index path. NAME ONLY -- git refuses an existing zero-byte index
/// file, so we must not create it first.
fn temp_index_path() -> PathBuf {
    let mut p = std::env::temp_dir();
    p.push(format!(
        "jedimem-idx-{}-{}",
        std::process::id(),
        memory_nonce()
    ));
    p
}

fn memory_nonce() -> String {
    let mut b = [0u8; 6];
    let _ = getrandom::getrandom(&mut b);
    b.iter().map(|x| format!("{:02x}", x)).collect()
}

fn build_tree(files: &BTreeMap<String, String>, parent: &str, cwd: Option<&Path>) -> R<String> {
    let idx = temp_index_path();
    let idx_s = idx.to_string_lossy().into_owned();
    let result = (|| -> R<String> {
        if !parent.is_empty() {
            let out = base(cwd)
                .args(["read-tree", parent])
                .env("GIT_INDEX_FILE", &idx_s)
                .output()
                .map_err(|e| GitError(e.to_string()))?;
            if !out.status.success() {
                return Err(GitError(String::from_utf8_lossy(&out.stderr).into_owned()));
            }
        }
        for (path, blob) in files {
            let out = base(cwd)
                .args([
                    "update-index",
                    "--add",
                    "--cacheinfo",
                    &format!("100644,{},{}", blob, path),
                ])
                .env("GIT_INDEX_FILE", &idx_s)
                .output()
                .map_err(|e| GitError(e.to_string()))?;
            if !out.status.success() {
                return Err(GitError(String::from_utf8_lossy(&out.stderr).into_owned()));
            }
        }
        let out = base(cwd)
            .arg("write-tree")
            .env("GIT_INDEX_FILE", &idx_s)
            .output()
            .map_err(|e| GitError(e.to_string()))?;
        if !out.status.success() {
            return Err(GitError(String::from_utf8_lossy(&out.stderr).into_owned()));
        }
        Ok(String::from_utf8_lossy(&out.stdout).trim().to_string())
    })();
    let _ = std::fs::remove_file(&idx);
    result
}

fn empty_tree(cwd: Option<&Path>) -> R<String> {
    check(&["hash-object", "-w", "-t", "tree", "--stdin"], cwd)
        .or_else(|_| Ok("4b825dc642cb6eb9a060e54bf8d69288fbee4904".to_string()))
}

/// Commit `files` onto a side ref without touching index/HEAD/worktree.
///
/// Compare-and-swap on the ref, retrying on contention. The parent MUST be read
/// before the tree is built: doing it the other way round passes the CAS while
/// committing a stale tree, which silently drops other writers' work.
pub fn stage_files(
    files: &BTreeMap<String, String>,
    message: &str,
    r: &str,
    cwd: Option<&Path>,
) -> R<String> {
    let mut blobs = BTreeMap::new();
    for (p, c) in files {
        blobs.insert(p.clone(), write_blob(c, cwd)?);
    }
    for _ in 0..100 {
        let parent = ref_exists(r, cwd);
        let tree = build_tree(&blobs, &parent, cwd)?;
        let commit = if parent.is_empty() {
            check(&["commit-tree", &tree, "-m", message], cwd)?
        } else {
            check(&["commit-tree", &tree, "-p", &parent, "-m", message], cwd)?
        };
        let (_, ok) = run(&["update-ref", r, &commit, &parent], cwd)?;
        if ok {
            return Ok(commit);
        }
    }
    Err(GitError(format!(
        "could not update {} after 100 attempts",
        r
    )))
}

pub fn staged_files(r: &str, cwd: Option<&Path>, prefix: &str) -> Vec<String> {
    if ref_exists(r, cwd).is_empty() {
        return vec![];
    }
    match run(&["ls-tree", "-r", "--name-only", r], cwd) {
        Ok((out, true)) => out
            .lines()
            .filter(|p| p.starts_with(prefix))
            .map(String::from)
            .collect(),
        _ => vec![],
    }
}

pub fn staged_content(path: &str, r: &str, cwd: Option<&Path>) -> R<String> {
    check(&["show", &format!("{}:{}", r, path)], cwd)
}

/// Rewrite the staging ref without `drop` (after promote/reject).
pub fn drop_staged(drop: &[String], r: &str, cwd: Option<&Path>, message: &str) -> R<String> {
    let mut keep: BTreeMap<String, String> = BTreeMap::new();
    for p in staged_files(r, cwd, "") {
        if !drop.contains(&p) {
            let c = staged_content(&p, r, cwd)?;
            keep.insert(p, format!("{}\n", c));
        }
    }
    let mut blobs = BTreeMap::new();
    for (p, c) in &keep {
        blobs.insert(p.clone(), write_blob(c, cwd)?);
    }
    let parent = ref_exists(r, cwd);
    let tree = if blobs.is_empty() {
        empty_tree(cwd)?
    } else {
        build_tree(&blobs, "", cwd)?
    };
    let commit = if parent.is_empty() {
        check(&["commit-tree", &tree, "-m", message], cwd)?
    } else {
        check(&["commit-tree", &tree, "-p", &parent, "-m", message], cwd)?
    };
    check(&["update-ref", r, &commit, &parent], cwd)?;
    Ok(commit)
}
